File: lib_book_common/src/main/java/com/ebook/common/view/profilePhoto/PhotoCutDialog.java
package com.ebook.common.view.profilePhoto;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.ebook.common.R;
import com.ebook.common.util.MultiMediaUtil;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.List;

import me.nereo.multi_image_selector.MultiImageSelectorActivity;


public class PhotoCutDialog extends BottomSheetDialogFragment implements View.OnClickListener {

    public static final String TAG = PhotoCutDialog.class.getSimpleName();
    //请求截图
    private static final int REQUEST_CROP_PHOTO = 102;
    private OnPhotoClickListener mOnClickListener;
    private String mPhotoPath;

    public static PhotoCutDialog newInstance() {
        return new PhotoCutDialog();
    }

    public void setOnClickListener(OnPhotoClickListener onPhotoClickListener) {
        mOnClickListener = onPhotoClickListener;
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_photo_select, container, false);
        Button btnSelectPhoto = view.findViewById(R.id.btn_select_photo);
        Button btnTakePhoto = view.findViewById(R.id.btn_take_photo);
        Button btnCancel = view.findViewById(R.id.btn_cancel);

        btnSelectPhoto.setOnClickListener(this);
        btnTakePhoto.setOnClickListener(this);
        btnCancel.setOnClickListener(this);
        return view;
    }

    @Override
    public void onClick(View v) {
        int i = v.getId();
        if (i == R.id.btn_take_photo) {
            mPhotoPath = MultiMediaUtil.getPhotoPath((AppCompatActivity) requireActivity());
            // Log.e(TAG, "onClick: 照相:  "+mPhotoPath);
            MultiMediaUtil.takePhoto(this, mPhotoPath, MultiMediaUtil.TAKE_PHONE);

        } else if (i == R.id.btn_select_photo) {
            MultiMediaUtil.pohotoSelect(this, 1, MultiMediaUtil.SELECT_IMAGE);

        } else if (i == R.id.btn_cancel) {
            dismiss();

        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case MultiMediaUtil.SELECT_IMAGE:
                if (resultCode == Activity.RESULT_OK) {
                    List<String> path = data.getStringArrayListExtra(MultiImageSelectorActivity.EXTRA_RESULT);
                    if (path != null) {
                        Uri uri = Uri.parse(path.get(0));
                        gotoClipActivity(uri);
                    }

                }
                break;
            case MultiMediaUtil.TAKE_PHONE:
                if (resultCode == Activity.RESULT_OK) {
                    Uri uri = Uri.parse(mPhotoPath);
                    gotoClipActivity(uri);
                }
                break;
            case REQUEST_CROP_PHOTO:
                if (resultCode == Activity.RESULT_OK) {
                    final Uri uri = data.getData();
                    if (mOnClickListener != null) {
                        mOnClickListener.onScreenPhotoClick(uri);
                    }
                    dismiss();
                }

                break;
        }
    }

    /**
     * 打开截图界面
     */
    private void gotoClipActivity(Uri uri) {
        if (uri == null) {
            return;
        }
        Intent intent = new Intent();
        intent.setClass(requireActivity(), ClipImageActivity.class);
        // 1: 圆形 2: 正方形
        int cutType = 1;
        intent.putExtra("type", cutType);
        intent.setData(uri);
        startActivityForResult(intent, REQUEST_CROP_PHOTO);
    }

    public interface OnPhotoClickListener {
        void onScreenPhotoClick(Uri uri);

    }
}

File: lib_book_common/src/main/java/com/ebook/common/view/profilePhoto/PhotoCutDialog.kt
package com.ebook.common.view.profilePhoto

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import com.ebook.common.databinding.FragmentPhotoSelectBinding
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.xrn1997.common.util.FileUtil
import java.io.File

class PhotoCutDialog : BottomSheetDialogFragment() {
    private var mOnClickListener: ((uri: Uri) -> Unit)? = null
    private lateinit var mPhotoFile: File

    // 注册用于接收 activity 结果的启动器
    private lateinit var selectLauncher: ActivityResultLauncher<PickVisualMediaRequest>
    private lateinit var takePhotoLauncher: ActivityResultLauncher<Uri>
    private lateinit var cropPhotoLauncher: ActivityResultLauncher<Intent>
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPhotoFile = FileUtil.getPrivateFile(requireContext(), "profile.jpeg")
        selectLauncher =
            registerForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
                if (uri != null) {
                    gotoClipActivity(uri)
                }
            }

        takePhotoLauncher =
            registerForActivityResult(ActivityResultContracts.TakePicture()) { success ->
                if (success) {
                    gotoClipActivity(Uri.fromFile(mPhotoFile))
                }
            }

        cropPhotoLauncher =
            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
                if (result.resultCode == Activity.RESULT_OK) {
                    result.data?.data?.let {
                        mOnClickListener?.invoke(it)
                    }
                    dismiss()
                }
            }
    }

    fun setOnClickListener(onPhotoClickListener: (uri: Uri) -> Unit) {
        mOnClickListener = onPhotoClickListener
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = FragmentPhotoSelectBinding.inflate(inflater, container, false)
        binding.btnSelectPhoto.setOnClickListener {
            selectLauncher.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly))
        }
        binding.btnTakePhoto.setOnClickListener {
            takePhotoLauncher.launch(FileUtil.getUri(requireContext(), mPhotoFile))
        }
        binding.btnCancel.setOnClickListener { dismiss() }
        return binding.root
    }


    /**
     * 打开截图界面
     */
    private fun gotoClipActivity(uri: Uri?) {
        if (uri == null) return
        val intent = Intent()
        intent.setClass(requireActivity(), ClipImageActivity::class.java)
        // 1: 圆形 2: 正方形
        val cutType = 1
        intent.putExtra("type", cutType)
        intent.setData(uri)
        cropPhotoLauncher.launch(intent)
    }

    companion object {
        private val TAG: String = PhotoCutDialog::class.java.simpleName

        @JvmStatic
        fun newInstance(): PhotoCutDialog {
            return PhotoCutDialog()
        }
    }
}
